File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <bits/stdc++.h>

using namespace std;

int backtracking(vector<int> numbers, int target, int index, int sum) {
    if(index == numbers.size()) {
        return target == sum;
    }
    
    return backtracking(numbers, target, index + 1, sum + numbers[index])
        + backtracking(numbers, target, index + 1, sum - numbers[index]);
}

int solution(vector<int> numbers, int target) {
    return backtracking(numbers, target, 0, 0);
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int dfs(int[] numbers, int idx, int sum, int target) {
        if(idx == numbers.length) {
            return (sum == target) ? 1 : 0;
        }

        return dfs(numbers, idx+1, sum+numbers[idx], target) + 
            dfs(numbers, idx+1, sum-numbers[idx], target);
    }


    public int solution(int[] numbers, int target) {
        return dfs(numbers, 0, 0, target);
    }
}